sunscreen_runtime/src/runtime.rs

```diff
@@ -20,21 +22,77 @@ enum Context {
 }
 
 /**
- * Contains all the elements needed to encrypt, decrypt, generate keys, and evaluate FHE programs.
+ * Marker traits.
  */
-pub struct Runtime {
+pub mod marker {
     /**
-     * The parameters used to construct the scheme used in this runtime.
+     * A marker trait that denotes a [`Runtime`] can perform FHE operations.
      */
-    params: Params,
+    pub trait Fhe {}
 
     /**
-     * The context associated with the BFV scheme.
+     * A marker trait that denotes a [`Runtime`] can perform ZKP operations.
      */
+    pub trait Zkp {}
+}
+
+/**
+ * A surrogate type for creating FHE-enabled [`Runtime`]s.
+ */
+pub struct Fhe {}
+impl marker::Fhe for Fhe {}
+
+/**
+ * A surrogate type for creating ZKP-enabled [`Runtime`]s.
+ */
+pub struct Zkp {}
+impl marker::Zkp for Zkp {}
+
+/**
+ * A surrogate type for creating [`Runtime`]s supporting both FHE and ZKP.
+ */
+pub struct FheZkp {}
+impl marker::Fhe for FheZkp {}
+impl marker::Zkp for FheZkp {}
+
+struct FheRuntimeData {
+    params: Params,
     context: Context,
 }
 
-impl Runtime {
+enum RuntimeData {
+    Fhe(FheRuntimeData),
+    Zkp,
+}
+
+impl RuntimeData {
+    /**
+     * Gets the inner Fhe's runtime data or panics if this value isn't
+     * the [`RuntimeData::Fhe`] variant.
+     *
+     * # Panics
+     * * If this value isn't a [`RuntimeData::Fhe`].
+     */
+    fn unwrap_fhe(&self) -> &FheRuntimeData {
+        match self {
+            Self::Fhe(x) => x,
+            _ => panic!("Expected RuntimeData::Fhe."),
+        }
+    }
+}
+
+/**
+ * Contains all the elements needed to encrypt, decrypt, generate keys, and evaluate FHE programs.
```

Comment:
This comment indicates the type is specific to FHE programs, but the type looks more general than that.